Documentation/Examples/RetriableSample.dib

```diff
@@ -0,0 +1,660 @@
+
+#!markdown
+
+#### `Backoff`
+This retries a specified number of times, with a delay that increases geometrically between each retry.
+
+For example, `new Backoff(maxTries: 5, deltaBackoff: TimeSpan.FromSeconds(1))` will retry up to 5 times. Each time it will increase the delay by a value calculated roughly like this: `2^n * (delta +/- a small random fudge)`, where `n` is the current number of retries.
+
+This allows a rapidly increasing delay, with a bit of random jitter addded to avoid contention.
+
+You can also set a `MinBackoff` and `MaxBackoff` to limit the lower and upper bounds of the delay.
+
+##### Example
+
+#!csharp
+
+MockHttpService mockHttpService = new();
+Clock clock = new();
+Backoff backoffStrategy = new(maxTries: 4, deltaBackoff: TimeSpan.FromSeconds(1));
+Retriable.Retry(() => mockHttpService.MakeLotsOfRequests(clock), CancellationToken.None, strategy: backoffStrategy, policy: new RetryOnTooManyRequestsPolicy());
+
+#!markdown
+
+##### Example - using `MinBackoff` and `MaxBackoff`
+
+`MinBackoff` and `MaxBackoff` are available as settable properties on a `BackOff` object. If we use the same example as above but set the `MaxBackoff` to 5 seconds, we will see that once the number of retries reaches a number such that  delay > 5 seconds, the delay between each retry will be restricted to 5 seconds.
+
+#!csharp
+
+MockHttpService mockHttpService = new();
+Clock clock = new();
+Backoff backoffStrategy = new(maxTries: 5, deltaBackoff: TimeSpan.FromSeconds(1));
+backoffStrategy.MaxBackoff = TimeSpan.FromSeconds(5);
+Retriable.Retry(() => mockHttpService.MakeLotsOfRequests(clock), CancellationToken.None, strategy: backoffStrategy, policy: new RetryOnTooManyRequestsPolicy());
+
+#!markdown
+
+And, if we set the `MinBackoff` to be 2 seconds (the default is 1 second), we will see the time elapsed between each retry increase by 1 second on average.
+
+#!csharp
+
+MockHttpService mockHttpService = new();
+Clock clock = new();
+Backoff backoffStrategy = new(maxTries: 4, deltaBackoff: TimeSpan.FromSeconds(1));
+backoffStrategy.MinBackoff = TimeSpan.FromSeconds(2);
+Retriable.Retry(() => mockHttpService.MakeLotsOfRequests(clock), CancellationToken.None, strategy: backoffStrategy, policy: new RetryOnTooManyRequestsPolicy());
+
+#!markdown
+
+#### Implementing a custom strategy
+
+It is slightly more complex to implement a custom retry strategy than it was to implement a retry policy. Although you can directly implemet the `IRetryStrategy` interface, it is usually simpler to derive from the abstract `RetryStrategy` class.
+
+In that case, you need to override the `TimeSpan PrepareToRetry(Exception lastException)` method. You can examine the exception and use your internal state to determine two things:
+
+1) Should we record this exception, or filter it out? Typically, we will want to record the exception, and, if so, we call `this.AddException(lastException)`.
```

Comment:
It's not just "should we record it" but also "do we want to create a new exception (with the original as an inner exception) which decorates it with more information.

Reply:
I'm not sure how this would be used. Can you give an example?

Reply:
I think instead of doing this:

```cs
this.AddException(lastException);
```

in your `PrepareToRetry`,  you could do this:

```cs
this.AddException(new MyWrapperException(lastException));
```

And...I guess we could in principle report stuff like how many retries were made, or how long we retried for.

That said, I can't think of an example where we've actually done this.

Reply:
Yes - it is one of those "this is the kind of thing you might want to do". The fact we *haven't* wanted to is interesting; but the choice is yours!

Reply:
I've added an example that wraps `lastException` with a new exception with a property holding the `tryCount` value that caused the particular exception to be thrown, and records that exception.